README
= AWS::S3

AWS::S3 is a Ruby library for Amazon's Simple Storage Service's REST API (http://aws.amazon.com/s3).
Full documentation of the currently supported API can be found at http://docs.amazonwebservices.com/AmazonS3/2006-03-01.

== Getting started

To get started you need to require 'aws/s3':

  % irb -rubygems
  irb(main):001:0> require 'aws/s3'
  # => true

The AWS::S3 library ships with an interactive shell called <tt>s3sh</tt>. From within it, you have access to all the operations the library exposes from the command line.

  % s3sh
  >> Version

Before you can do anything, you must establish a connection using Base.establish_connection!.  A basic connection would look something like this:

  AWS::S3::Base.establish_connection!(
    :access_key_id     => 'abc', 
    :secret_access_key => '123'
  )

The minimum connection options that you must specify are your access key id and your secret access key.

(If you don't already have your access keys, all you need to sign up for the S3 service is an account at Amazon. You can sign up for S3 and get access keys by visiting http://aws.amazon.com/s3.)

For convenience, if you set two special environment variables with the value of your access keys, the console will automatically create a default connection for you. For example:

  % cat .amazon_keys
  export AMAZON_ACCESS_KEY_ID='abcdefghijklmnop'
  export AMAZON_SECRET_ACCESS_KEY='1234567891012345'

Then load it in your shell's rc file.

  % cat .zshrc
  if [[ -f "$HOME/.amazon_keys" ]]; then
    source "$HOME/.amazon_keys";
  fi

See more connection details at AWS::S3::Connection::Management::ClassMethods.


== AWS::S3 Basics
=== The service, buckets and objects

The three main concepts of S3 are the service, buckets and objects. 

==== The service

The service lets you find out general information about your account, like what buckets you have. 

  Service.buckets
  # => []

 
==== Buckets

Buckets are containers for objects (the files you store on S3). To create a new bucket you just specify its name.

  # Pick a unique name, or else you'll get an error 
  # if the name is already taken.
  Bucket.create('jukebox')

Bucket names must be unique across the entire S3 system, sort of like domain names across the internet. If you try
to create a bucket with a name that is already taken, you will get an error.

Assuming the name you chose isn't already taken, your new bucket will now appear in the bucket list:

  Service.buckets
  # => [#<AWS::S3::Bucket @attributes={"name"=>"jukebox"}>]

Once you have succesfully created a bucket you can you can fetch it by name using Bucket.find.

  music_bucket = Bucket.find('jukebox')

The bucket that is returned will contain a listing of all the objects in the bucket.

  music_bucket.objects.size
  # => 0

If all you are interested in is the objects of the bucket, you can get to them directly using Bucket.objects.

  Bucket.objects('jukebox').size
  # => 0

By default all objects will be returned, though there are several options you can use to limit what is returned, such as
specifying that only objects whose name is after a certain place in the alphabet be returned, and etc. Details about these options can
be found in the documentation for Bucket.find.

To add an object to a bucket you specify the name of the object, its value, and the bucket to put it in.

  file = 'black-flowers.mp3'
  S3Object.store(file, open(file), 'jukebox')

You'll see your file has been added to it:

  music_bucket.objects
  # => [#<AWS::S3::S3Object '/jukebox/black-flowers.mp3'>]

You can treat your bucket like a hash and access objects by name:

  jukebox['black-flowers.mp3']
  # => #<AWS::S3::S3Object '/jukebox/black-flowers.mp3'>

In the event that you want to delete a bucket, you can use Bucket.delete.

  Bucket.delete('jukebox')

Keep in mind, like unix directories, you can not delete a bucket unless it is empty. Trying to delete a bucket
that contains objects will raise a BucketNotEmpty exception.

Passing the :force => true option to delete will take care of deleting all the bucket's objects for you.

  Bucket.delete('photos', :force => true)
  # => true


==== Objects

S3Objects represent the data you store on S3. They have a key (their name) and a value (their data). All objects belong to a
bucket.

You can store an object on S3 by specifying a key, its data and the name of the bucket you want to put it in:

  S3Object.store('me.jpg', open('headshot.jpg'), 'photos')

The content type of the object will be inferred by its extension. If the appropriate content type can not be inferred, S3 defaults
to <tt>binary/octet-stream</tt>.

If you want to override this, you can explicitly indicate what content type the object should have with the <tt>:content_type</tt> option:

  file = 'black-flowers.m4a'
  S3Object.store(
    file,
    open(file),
    'jukebox',
    :content_type => 'audio/mp4a-latm'
  )

You can read more about storing files on S3 in the documentation for S3Object.store.

If you just want to fetch an object you've stored on S3, you just specify its name and its bucket:

  picture = S3Object.find 'headshot.jpg', 'photos'

N.B. The actual data for the file is not downloaded in both the example where the file appeared in the bucket and when fetched directly. 
You get the data for the file like this:

  picture.value

You can fetch just the object's data directly:

  S3Object.value 'headshot.jpg', 'photos'

Or stream it by passing a block to <tt>stream</tt>:

  open('song.mp3', 'w') do |file|
    S3Object.stream('song.mp3', 'jukebox') do |chunk|
      file.write chunk
    end
  end

The data of the file, once download, is cached, so subsequent calls to <tt>value</tt> won't redownload the file unless you 
tell the object to reload its <tt>value</tt>:

  # Redownloads the file's data
  song.value(:reload) 

Other functionality includes:

  # Check if an object exists?
  S3Object.exists? 'headshot.jpg', 'photos'

  # Copying an object
  S3Object.copy 'headshot.jpg', 'headshot2.jpg', 'photos'

  # Renaming an object
  S3Object.rename 'headshot.jpg', 'portrait.jpg', 'photos'

  # Deleting an object
  S3Object.delete 'headshot.jpg', 'photos'

==== More about objects and their metadata

You can find out the content type of your object with the <tt>content_type</tt> method:

  song.content_type
  # => "audio/mpeg"

You can change the content type as well if you like:

  song.content_type = 'application/pdf'
  song.store

(Keep in mind that due to limitiations in S3's exposed API, the only way to change things like the content_type
is to PUT the object onto S3 again. In the case of large files, this will result in fully re-uploading the file.)

A bevie of information about an object can be had using the <tt>about</tt> method:

  pp song.about
  {"last-modified"    => "Sat, 28 Oct 2006 21:29:26 GMT",
   "content-type"     => "binary/octet-stream",
   "etag"             => "\"dc629038ffc674bee6f62eb64ff3a\"",
   "date"             => "Sat, 28 Oct 2006 21:30:41 GMT",
   "x-amz-request-id" => "B7BC68F55495B1C8",
   "server"           => "AmazonS3",
   "content-length"   => "3418766"}

You can get and set metadata for an object:

  song.metadata
  # => {}
  song.metadata[:album] = "A River Ain't Too Much To Love"
  # => "A River Ain't Too Much To Love"
  song.metadata[:released] = 2005
  pp song.metadata
  {"x-amz-meta-released" => 2005, 
    "x-amz-meta-album"   => "A River Ain't Too Much To Love"}
  song.store

That metadata will be saved in S3 and is hence forth available from that object:

  song = S3Object.find('black-flowers.mp3', 'jukebox')
  pp song.metadata
  {"x-amz-meta-released" => "2005", 
    "x-amz-meta-album"   => "A River Ain't Too Much To Love"}
  song.metada[:released]
  # => "2005"
  song.metada[:released] = 2006
  pp song.metada
  {"x-amz-meta-released" => 2006, 
   "x-amz-meta-album"    => "A River Ain't Too Much To Love"}


==== Streaming uploads

When storing an object on the S3 servers using S3Object.store, the <tt>data</tt> argument can be a string or an I/O stream. 
If <tt>data</tt> is an I/O stream it will be read in segments and written to the socket incrementally. This approach 
may be desirable for very large files so they are not read into memory all at once.

  # Non streamed upload
  S3Object.store('greeting.txt', 'hello world!', 'marcel')

  # Streamed upload
  S3Object.store('roots.mpeg', open('roots.mpeg'), 'marcel')


== Setting the current bucket
==== Scoping operations to a specific bucket

If you plan on always using a specific bucket for certain files, you can skip always having to specify the bucket by creating 
a subclass of Bucket or S3Object and telling it what bucket to use:

  class JukeBoxSong < AWS::S3::S3Object
    set_current_bucket_to 'jukebox'
  end

For all methods that take a bucket name as an argument, the current bucket will be used if the bucket name argument is omitted.

  other_song = 'baby-please-come-home.mp3'
  JukeBoxSong.store(other_song, open(other_song))

This time we didn't have to explicitly pass in the bucket name, as the JukeBoxSong class knows that it will
always use the 'jukebox' bucket. 

"Astute readers", as they say, may have noticed that we used the third parameter to pass in the content type,
rather than the fourth parameter as we had the last time we created an object. If the bucket can be inferred, or
is explicitly set, as we've done in the JukeBoxSong class, then the third argument can be used to pass in
options.

Now all operations that would have required a bucket name no longer do.

  other_song = JukeBoxSong.find('baby-please-come-home.mp3')


== BitTorrent
==== Another way to download large files

Objects on S3 can be distributed via the BitTorrent file sharing protocol. 

You can get a torrent file for an object by calling <tt>torrent_for</tt>:

  S3Object.torrent_for 'kiss.jpg', 'marcel'

Or just call the <tt>torrent</tt> method if you already have the object:

  song = S3Object.find 'kiss.jpg', 'marcel'
  song.torrent

Calling <tt>grant_torrent_access_to</tt> on a object will allow anyone to anonymously
fetch the torrent file for that object:

  S3Object.grant_torrent_access_to 'kiss.jpg', 'marcel'

Anonymous requests to

  http://s3.amazonaws.com/marcel/kiss.jpg?torrent

will serve up the torrent file for that object.


== Access control
==== Using canned access control policies

By default buckets are private. This means that only the owner has access rights to the bucket and its objects. 
Objects in that bucket inherit the permission of the bucket unless otherwise specified. When an object is private, the owner can 
generate a signed url that exposes the object to anyone who has that url. Alternatively, buckets and objects can be given other 
access levels. Several canned access levels are defined:

* <tt>:private</tt> - Owner gets FULL_CONTROL. No one else has any access rights. This is the default.
* <tt>:public_read</tt> - Owner gets FULL_CONTROL and the anonymous principal is granted READ access. If this policy is used on an object, it can be read from a browser with no authentication.
* <tt>:public_read_write</tt> - Owner gets FULL_CONTROL, the anonymous principal is granted READ and WRITE access. This is a useful policy to apply to a bucket, if you intend for any anonymous user to PUT objects into the bucket.
* <tt>:authenticated_read</tt> - Owner gets FULL_CONTROL, and any principal authenticated as a registered Amazon S3 user is granted READ access.

You can set a canned access level when you create a bucket or an object by using the <tt>:access</tt> option:

  S3Object.store(
    'kiss.jpg', 
    data, 
    'marcel', 
    :access => :public_read
  )

Since the image we created is publicly readable, we can access it directly from a browser by going to the corresponding bucket name 
and specifying the object's key without a special authenticated url:

 http://s3.amazonaws.com/marcel/kiss.jpg

==== Building custum access policies

For both buckets and objects, you can use the <tt>acl</tt> method to see its access control policy:

  policy = S3Object.acl('kiss.jpg', 'marcel')
  pp policy.grants
  [#<AWS::S3::ACL::Grant FULL_CONTROL to noradio>,
   #<AWS::S3::ACL::Grant READ to AllUsers Group>]

Policies are made up of one or more grants which grant a specific permission to some grantee. Here we see the default FULL_CONTROL grant 
to the owner of this object. There is also READ permission granted to the Allusers Group, which means anyone has read access for the object.

Say we wanted to grant access to anyone to read the access policy of this object. The current READ permission only grants them permission to read 
the object itself (for example, from a browser) but it does not allow them to read the access policy. For that we will need to grant the AllUsers group the READ_ACP permission.

First we'll create a new grant object:

  grant = ACL::Grant.new
  # => #<AWS::S3::ACL::Grant (permission) to (grantee)>
  grant.permission = 'READ_ACP'

Now we need to indicate who this grant is for. In other words, who the grantee is:

  grantee = ACL::Grantee.new
  # => #<AWS::S3::ACL::Grantee (xsi not set yet)>

There are three ways to specify a grantee: 1) by their internal amazon id, such as the one returned with an object's Owner, 
2) by their Amazon account email address or 3) by specifying a group. As of this writing you can not create custom groups, but 
Amazon does provide three already: AllUsers, Authenticated and LogDelivery. In this case we want to provide the grant to all users. 
This effectively means "anyone".

  grantee.group = 'AllUsers'

Now that our grantee is setup, we'll associate it with the grant:

  grant.grantee = grantee
  grant
  # => #<AWS::S3::ACL::Grant READ_ACP to AllUsers Group>

Are grant has all the information we need. Now that it's ready, we'll add it on to the object's access control policy's list of grants:

  policy.grants << grant
  pp policy.grants
  [#<AWS::S3::ACL::Grant FULL_CONTROL to noradio>,
   #<AWS::S3::ACL::Grant READ to AllUsers Group>,
   #<AWS::S3::ACL::Grant READ_ACP to AllUsers Group>]

Now that the policy has the new grant, we reuse the <tt>acl</tt> method to persist the policy change:

  S3Object.acl('kiss.jpg', 'marcel', policy)

If we fetch the object's policy again, we see that the grant has been added:

  pp S3Object.acl('kiss.jpg', 'marcel').grants
  [#<AWS::S3::ACL::Grant FULL_CONTROL to noradio>,
   #<AWS::S3::ACL::Grant READ to AllUsers Group>,
   #<AWS::S3::ACL::Grant READ_ACP to AllUsers Group>]

If we were to access this object's acl url from a browser: 

  http://s3.amazonaws.com/marcel/kiss.jpg?acl

we would be shown its access control policy.

==== Pre-prepared grants

Alternatively, the ACL::Grant class defines a set of stock grant policies that you can fetch by name. In most cases, you can 
just use one of these pre-prepared grants rather than building grants by hand. Two of these stock policies are <tt>:public_read</tt> 
and <tt>:public_read_acp</tt>, which happen to be the two grants that we built by hand above. In this case we could have simply written:

  policy.grants << ACL::Grant.grant(:public_read)
  policy.grants << ACL::Grant.grant(:public_read_acp)
  S3Object.acl('kiss.jpg', 'marcel', policy)

The full details can be found in ACL::Policy, ACL::Grant and ACL::Grantee.


==== Accessing private objects from a browser

All private objects are accessible via an authenticated GET request to the S3 servers. You can generate an 
authenticated url for an object like this:

  S3Object.url_for('beluga_baby.jpg', 'marcel_molina')

By default authenticated urls expire 5 minutes after they were generated.

Expiration options can be specified either with an absolute time since the epoch with the <tt>:expires</tt> options,
or with a number of seconds relative to now with the <tt>:expires_in</tt> options:

  # Absolute expiration date 
  # (Expires January 18th, 2038)
  doomsday = Time.mktime(2038, 1, 18).to_i
  S3Object.url_for('beluga_baby.jpg', 
                   'marcel', 
                   :expires => doomsday)
  
  # Expiration relative to now specified in seconds 
  # (Expires in 3 hours)
  S3Object.url_for('beluga_baby.jpg', 
                   'marcel', 
                   :expires_in => 60 * 60 * 3)

You can specify whether the url should go over SSL with the <tt>:use_ssl</tt> option:

  # Url will use https protocol
  S3Object.url_for('beluga_baby.jpg', 
                   'marcel', 
                   :use_ssl => true)

By default, the ssl settings for the current connection will be used.

If you have an object handy, you can use its <tt>url</tt> method with the same objects:

  song.url(:expires_in => 30)

To get an unauthenticated url for the object, such as in the case
when the object is publicly readable, pass the
<tt>:authenticated</tt> option with a value of <tt>false</tt>.

  S3Object.url_for('beluga_baby.jpg',
                   'marcel',
                   :authenticated => false)
  # => http://s3.amazonaws.com/marcel/beluga_baby.jpg


== Logging
==== Tracking requests made on a bucket

A bucket can be set to log the requests made on it. By default logging is turned off. You can check if a bucket has logging enabled:

  Bucket.logging_enabled_for? 'jukebox'
  # => false

Enabling it is easy:

  Bucket.enable_logging_for('jukebox')

Unless you specify otherwise, logs will be written to the bucket you want to log. The logs are just like any other object. By default they will start with the prefix 'log-'. You can customize what bucket you want the logs to be delivered to, as well as customize what the log objects' key is prefixed with by setting the <tt>target_bucket</tt> and <tt>target_prefix</tt> option:

  Bucket.enable_logging_for(
    'jukebox', 'target_bucket' => 'jukebox-logs'
  )

Now instead of logging right into the jukebox bucket, the logs will go into the bucket called jukebox-logs.

Once logs have accumulated, you can access them using the <tt>logs</tt> method:

  pp Bucket.logs('jukebox')
  [#<AWS::S3::Logging::Log '/jukebox-logs/log-2006-11-14-07-15-24-2061C35880A310A1'>,
   #<AWS::S3::Logging::Log '/jukebox-logs/log-2006-11-14-08-15-27-D8EEF536EC09E6B3'>,
   #<AWS::S3::Logging::Log '/jukebox-logs/log-2006-11-14-08-15-29-355812B2B15BD789'>]


CHANGELOG
head:

0.6.2:

- Apparently need to use custom __method__ in Ruby versions right up to 1.8.7.

0.6.1:

- Use custom __method__ in Ruby versions *prior* to 1.8.7 not *up to* 1.8.7.

- Rename Kernel#memoize to Kernel#expirable_memoize so that it doesn't conflict with memoize method in ActiveSupport which has an entirely different API and semantics. Reported by [Florian Dütsc (mail@florian-duetsch.de)].

0.6.0:

- Full 1.9 compatibility (all tests passing against 1.9 & 1.8.6). Thanks to [David (dvdplm@gmail.com), Cyril David (cyx.ucron@gmail.com)]

0.5.1:

- For now just redefine __method__ to take arguments so we don't break 1.8.7 use today

0.5.0:

- Bug #17458 fixed. Normalize the :expires_in option to always be an integer even if the actual object passed in is a proxy to an integer, such as is the case with 2.hours from ActiveSupport which is actually an instance of ActiveSupport::Duration. Reported by [Steve Kickert steve@riverocktech.com]

- Bug #19158 fixed. Don't prepend leading slash onto bucket name when deleting a bucket with the :force => true option.

- Bug #17628 fixed. Don't ignore :use_ssl => false in url_for when the connection is established over ssl. Reported by [Tom Fixed (tkat11)]

- Bug #13052 fixed. Appease some proxies by always including a Content-Length with all requests. Reported by [James Murty (jmurty)]

- Bug #13756 fixed. Attributes that are false should not raise NoMethodError in Base#method_missing. Fixed by [Scott Patten]

- Bug #19189 fixed. No longer reference Date::ABBR_MONTHS constant which was removed in Ruby 1.8.6. Reported by [Khurram Virani (kvirani)]

- Bug #20487 fixed. If a request fails and is retried, only escape the request path the first time. Reported by [anonymous]

- Replace ad-hoc S3Object.copy method with newly support built in API call.

- Do not make connections persistent by default. This "feature" causes far more broken pipes than it is worth. Use with caution.

0.4.0:

- Various adjustments to connection handling to try to mitigate exceptions raised from deep within Net::HTTP.

- Don't coerce numbers that start with a zero because the zero will be lost. If a bucket, for example, has a name like '0815', all operation it will fail. Closes ticket #10089 [reported anonymously]"

- Add ability to connect through a proxy using the :proxy option when establishing a connection. Suggested by [Simon Horne <simon@soulware.co.uk>]

- Add :authenticated option to url_for. When passing false, don't generate signature parameters for query string.

- Make url_for accept custom port settings. [Rich Olson] 

0.3.0:

- Ensure content type is eventually set to account for changes made to Net::HTTP in Ruby version 1.8.5. Reported by [David Hanson, Stephen Caudill, Tom Mornini <tmornini@engineyard.com>]

- Add :persistent option to connections which keeps a persistent connection rather than creating a new one per request, defaulting to true. Based on a patch by [Metalhead <metalhead@metalhead.ws>]

- If we are retrying a request after rescuing one of the retry exceptions, rewind the body if its an IO stream so it starts at the beginning. [Jamis Buck]

- Ensure that all paths being submitted to S3 are valid utf8. If they are not, we remove the extended characters. Ample help from [Jamis Buck]

- Wrap logs in Log objects which exposes each line as a Log::Line that has accessors by name for each field.

- Various performance optimizations for the extensions code. [Roman LE NEGRATE <roman2k@free.fr>]

- Make S3Object.copy more efficient by streaming in both directions in parallel.

- Open up Net:HTTPGenericRequest to make the chunk size 1 megabyte, up from 1 kilobyte.

- Add S3Object.exists?

0.2.1:

- When the bucket name argument (for e.g. Bucket.objects) is being used as the option hash, reassign it to the options variable and set the bucket to nil so bucket inference + options works.

- Don't call CGI.escape on query string parameters in Hash#to_query_string since all paths get passed through URI.escape right before the request is made. Paths were getting double escaped. Bug spotted by [David Hanson]

- Make s3sh exec irb.bat if on Windows. Bug spotted by [N. Sathish Kumar <nsathishk@yahoo.com>]

- Avoid class_variable_(get|set) since it was only recently added to Ruby. Spotted by [N. Sathish Kumar <nsathishk@yahoo.com>]

- Raise NoSuchKey if S3Object.about requests a key that does not exist.

- If the response body is an empty string, don't try to parse it as xml.

- Don't reject every body type save for IO and String at the door when making a request. Suggested by [Alex MacCaw <maccman@gmail.com>]

- Allow dots in bucket names. [Jesse Newland]

0.2.0:

- Infer content type for an object when calling S3Object.store without explicitly passing in the :content_type option.

0.1.2:

- Scrap (overly) fancy generator based version of CoercibleString with a much simpler and clearer case statement. Continuations are really slow and the specific use of the generator was leaking memory. Bug spotted by [Remco van't Veer]

0.1.1:

- Don't add the underscore method to String if it is already defined (like, for example, from ActiveSupport). Bug spotted by [Matt White <stockliasteroid@gmail.com>]

0.1.0:

- Initial public release


lib/aws/s3/logging.rb
module AWS
  module S3
    # A bucket can be set to log the requests made on it. By default logging is turned off. You can check if a bucket has logging enabled:
    # 
    #   Bucket.logging_enabled_for? 'jukebox'
    #   # => false
    # 
    # Enabling it is easy:
    # 
    #   Bucket.enable_logging_for('jukebox')
    # 
    # Unless you specify otherwise, logs will be written to the bucket you want to log. The logs are just like any other object. By default they will start with the prefix 'log-'. You can customize what bucket you want the logs to be delivered to, as well as customize what the log objects' key is prefixed with by setting the <tt>target_bucket</tt> and <tt>target_prefix</tt> option:
    # 
    #   Bucket.enable_logging_for(
    #     'jukebox', 'target_bucket' => 'jukebox-logs'
    #   )
    # 
    # Now instead of logging right into the jukebox bucket, the logs will go into the bucket called jukebox-logs.
    # 
    # Once logs have accumulated, you can access them using the <tt>logs</tt> method:
    # 
    #   pp Bucket.logs('jukebox')
    #   [#<AWS::S3::Logging::Log '/jukebox-logs/log-2006-11-14-07-15-24-2061C35880A310A1'>,
    #    #<AWS::S3::Logging::Log '/jukebox-logs/log-2006-11-14-08-15-27-D8EEF536EC09E6B3'>,
    #    #<AWS::S3::Logging::Log '/jukebox-logs/log-2006-11-14-08-15-29-355812B2B15BD789'>]
    #
    # Each log has a <tt>lines</tt> method that gives you information about each request in that log. All the fields are available 
    # as named methods. More information is available in Logging::Log::Line.
    #
    #   logs = Bucket.logs('jukebox')
    #   log  = logs.first
    #   line = log.lines.first
    #   line.operation
    #   # => 'REST.GET.LOGGING_STATUS'
    #   line.request_uri
    #   # => 'GET /jukebox?logging HTTP/1.1'
    #   line.remote_ip
    #   # => "67.165.183.125"
    #
    # Disabling logging is just as simple as enabling it:
    #
    #  Bucket.disable_logging_for('jukebox')
    module Logging
      # Logging status captures information about the calling bucket's logging settings. If logging is enabled for the bucket
      # the status object will indicate what bucket the logs are written to via the <tt>target_bucket</tt> method as well as
      # the logging key prefix with via <tt>target_prefix</tt>.
      #
      # See the documentation for Logging::Management::ClassMethods for more information on how to get the logging status of a bucket. 
      class Status
        include SelectiveAttributeProxy
        attr_reader :enabled
        alias_method :logging_enabled?, :enabled
      
        def initialize(attributes = {}) #:nodoc:
          attributes  = {'target_bucket' => nil, 'target_prefix' => nil}.merge(attributes)
          @enabled    = attributes.has_key?('logging_enabled')
          @attributes = attributes.delete('logging_enabled') || attributes
        end
      
        def to_xml #:nodoc:
          Builder.new(self).to_s
        end
      
        private
          attr_reader :attributes
        
          class Builder < XmlGenerator #:nodoc:
            attr_reader :logging_status
            def initialize(logging_status)
              @logging_status = logging_status
              super()
            end
          
            def build
              xml.tag!('BucketLoggingStatus', 'xmlns' => 'http://s3.amazonaws.com/doc/2006-03-01/') do
                if logging_status.target_bucket && logging_status.target_prefix
                  xml.LoggingEnabled do
                    xml.TargetBucket logging_status.target_bucket
                    xml.TargetPrefix logging_status.target_prefix
                  end
                end
              end
            end
          end
        end
      
      # A bucket log exposes requests made on the given bucket. Lines of the log represent a single request. The lines of a log
      # can be accessed with the lines method. 
      #
      #   log = Bucket.logs_for('marcel').first
      #   log.lines
      # 
      # More information about the logged requests can be found in the documentation for Log::Line.
      class Log
        def initialize(log_object) #:nodoc:
          @log = log_object
        end
        
        # Returns the lines for the log. Each line is wrapped in a Log::Line.
        def lines
          log.value.map {|line| Line.new(line)}
        end
        memoized :lines
        
        def path
          log.path
        end
        
        def inspect #:nodoc:
          "#<%s:0x%s '%s'>" % [self.class.name, object_id, path]
        end
        
        private
          attr_reader :log
          
        # Each line of a log exposes the raw line, but it also has method accessors for all the fields of the logged request.
        #
        # The list of supported log line fields are listed in the S3 documentation: http://docs.amazonwebservices.com/AmazonS3/2006-03-01/LogFormat.html
        #
        #   line = log.lines.first
        #   line.remote_ip
        #   # => '72.21.206.5'
        #
        # If a certain field does not apply to a given request (for example, the <tt>key</tt> field does not apply to a bucket request),
        # or if it was unknown or unavailable, it will return <tt>nil</tt>.
        #
        #   line.operation
        #   # => 'REST.GET.BUCKET'
        #   line.key
        #   # => nil
        class Line < String
          DATE          = /\[([^\]]+)\]/
          QUOTED_STRING = /"([^"]+)"/
          REST          = /(\S+)/
          LINE_SCANNER  = /#{DATE}|#{QUOTED_STRING}|#{REST}/
          
          cattr_accessor :decorators
          @@decorators = Hash.new {|hash, key| hash[key] = lambda {|entry| CoercibleString.coerce(entry)}}
          cattr_reader   :fields
          @@fields     = []
          
          class << self
            def field(name, offset, type = nil, &block) #:nodoc:
              decorators[name] = block if block_given?
              fields << name
              class_eval(<<-EVAL, __FILE__, __LINE__)
                def #{name}
                  value = parts[#{offset} - 1]
                  if value == '-'
                    nil
                  else
                    self.class.decorators[:#{name}].call(value)
                  end
                end
                memoized :#{name}
              EVAL
            end
            
            # Time.parse doesn't like %d/%B/%Y:%H:%M:%S %z so we have to transform it unfortunately
            def typecast_time(datetime) #:nodoc:
              month       = datetime[/[a-z]+/i]
              month_names = [nil, "Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]
              datetime.sub!(%r|^(\w{2})/(\w{3})|, '\2/\1')
              datetime.sub!(month, month_names.index(month).to_s)
              datetime.sub!(':', ' ')
              Time.parse(datetime)
            end
          end
          
          def initialize(line) #:nodoc:
            super(line)
            @parts = parse
          end
          
          field(:owner,            1)  {|entry| Owner.new('id' => entry) }
          field :bucket,           2
          field(:time,             3)  {|entry| typecast_time(entry)}
          field :remote_ip,        4
          field(:requestor,        5)  {|entry| Owner.new('id' => entry) }
          field :request_id,       6
          field :operation,        7
          field :key,              8
          field :request_uri,      9
          field :http_status,      10
          field :error_code,       11
          field :bytes_sent,       12
          field :object_size,      13
          field :total_time,       14
          field :turn_around_time, 15
          field :referrer,         16
          field :user_agent,       17
          
          # Returns all fields of the line in a hash of the form <tt>:field_name => :field_value</tt>.
          #
          #   line.attributes.values_at(:bucket, :key)
          #   # => ['marcel', 'kiss.jpg']
          def attributes
            self.class.fields.inject({}) do |attribute_hash, field|
              attribute_hash[field] = send(field)
              attribute_hash
            end
          end
          
          private
            attr_reader :parts
            
            def parse
              scan(LINE_SCANNER).flatten.compact
            end
        end
      end
      
      module Management #:nodoc:
        def self.included(klass) #:nodoc:
          klass.extend(ClassMethods)
          klass.extend(LoggingGrants)
        end
        
        module ClassMethods
          # Returns the logging status for the bucket named <tt>name</tt>. From the logging status you can determine the bucket logs are delivered to
          # and what the bucket object's keys are prefixed with. For more information see the Logging::Status class.
          #
          #   Bucket.logging_status_for 'marcel'
          def logging_status_for(name = nil, status = nil)
            if name.is_a?(Status)
              status = name
              name   = nil
            end

            path = path(name) << '?logging'
            status ? put(path, {}, status.to_xml) : Status.new(get(path).parsed)
          end
          alias_method :logging_status, :logging_status_for
          
          # Enables logging for the bucket named <tt>name</tt>. You can specify what bucket to log to with the <tt>'target_bucket'</tt> option as well
          # as what prefix to add to the log files with the <tt>'target_prefix'</tt> option. Unless you specify otherwise, logs will be delivered to
          # the same bucket that is being logged and will be prefixed with <tt>log-</tt>.
          def enable_logging_for(name = nil, options = {})
            name            = bucket_name(name)
            default_options = {'target_bucket' => name, 'target_prefix' => 'log-'}
            options         = default_options.merge(options)
            grant_logging_access_to_target_bucket(options['target_bucket'])
            logging_status(name, Status.new(options))
          end
          alias_method :enable_logging, :enable_logging_for
          
          # Disables logging for the bucket named <tt>name</tt>.
          def disable_logging_for(name = nil)
            logging_status(bucket_name(name), Status.new)
          end
          alias_method :disable_logging, :disable_logging_for
          
          # Returns true if logging has been enabled for the bucket named <tt>name</tt>.
          def logging_enabled_for?(name = nil)
            logging_status(bucket_name(name)).logging_enabled?
          end
          alias_method :logging_enabled?, :logging_enabled_for?
          
          # Returns the collection of logs for the bucket named <tt>name</tt>.
          #
          #   Bucket.logs_for 'marcel'
          #
          # Accepts the same options as Bucket.find, such as <tt>:max_keys</tt> and <tt>:marker</tt>.
          def logs_for(name = nil, options = {})
            if name.is_a?(Hash)
              options = name
              name    = nil
            end
            
            name           = bucket_name(name)
            logging_status = logging_status_for(name)
            return [] unless logging_status.logging_enabled?
            objects(logging_status.target_bucket, options.merge(:prefix => logging_status.target_prefix)).map do |log_object|
              Log.new(log_object)
            end
          end
          alias_method :logs, :logs_for
        end
        
        module LoggingGrants #:nodoc:
          def grant_logging_access_to_target_bucket(target_bucket)
            acl = acl(target_bucket)
            acl.grants << ACL::Grant.grant(:logging_write)
            acl.grants << ACL::Grant.grant(:logging_read_acp)
            acl(target_bucket, acl)
          end
        end
        
        def logging_status
          self.class.logging_status_for(name)
        end
        
        def enable_logging(*args)
          self.class.enable_logging_for(name, *args)
        end
        
        def disable_logging(*args)
          self.class.disable_logging_for(name, *args)
        end
        
        def logging_enabled?
          self.class.logging_enabled_for?(name)
        end
        
        def logs(options = {})
          self.class.logs_for(name, options)
        end
      end
    end
  end
end